src/control/system/membership.go

```diff
@@ -454,32 +454,62 @@ func (m *Membership) OnEvent(_ context.Context, evt events.Event) {
 		return
 	}
 
-	switch rankEvt := evt.(type) {
+	switch e := evt.(type) {
 	case *events.RankExit:
-		if rankEvt == nil {
-			m.log.Errorf("nil RankExit event received")
+		if e == nil {
+			m.log.Error("nil RankExit event received")
 			return
 		}
 		m.log.Debugf("processing RAS event %q from rank %d on host %q",
-			rankEvt.RAS.Msg, rankEvt.RAS.Rank, rankEvt.RAS.Hostname)
+			e.RAS.Msg, e.RAS.Rank, e.RAS.Hostname)
 
 		// TODO: sanity check that the correct member is being updated by
 		// performing lookup on provided hostname and matching returned
 		// addresses with the member address with matching rank.
 
 		if err := m.UpdateMemberStates(MemberResults{
-			NewMemberResult(Rank(rankEvt.RAS.Rank),
-				errors.Wrap(rankEvt.ExitErr, rankEvt.RAS.Msg),
+			NewMemberResult(Rank(e.RAS.Rank),
+				errors.Wrap(e.ExtendedInfo.ExitErr, e.RAS.Msg),
 				MemberStateErrored),
 		}, true); err != nil {
 			m.log.Errorf("updating member states: %s", err)
 			return
 		}
 
-		member, _ := m.Get(Rank(rankEvt.RAS.Rank))
-		m.log.Debugf("updated rank %d to %+v (%s)", rankEvt.RAS.Rank, member, member.Info)
+		member, _ := m.Get(Rank(e.RAS.Rank))
+		m.log.Debugf("update rank %d to %+v (%s)", e.RAS.Rank, member,
+			member.Info)
+	case *events.PoolSvcRanksUpdate:
```

Comment:
Seems a little weird to have the membership handling pool service updates. I get that it's convenient to just add this to the existing switch, but it might be prudent to step back a bit and rethink how these event handlers will grow over time. I can foresee new event handlers that don't fit into system.Membership and I'm concerned that setting this precedent now will add technical debt that we need to address later.

Reply:
could you give a suggestion as to how you would prefer this to be done, e.g. where the event handling switch should be instead

Reply:
Maybe a function in the server package that delegates based on the event id? It doesn't seem ideal to require that the `system.Membership` implementation should contain details about event handling.